americangut/diversity_analysis.py

```diff
@@ -0,0 +1,1729 @@
+            p_text.append('')
+
+    lines = []
+
+    # Plots the first comparison
+    l1 = []
+    l1.append(ax.plot([centers[0]]*2, [tops[0] + space, bar_levels[-1]], 'k-'))
+    l1.append(ax.plot([centers[1]]*2, [tops[1] + space, bar_levels[0]], 'k-'))
+    l1.append(ax.plot([centers[0], centers[1]], [bar_levels[0]]*2, 'k-'))
+    l1.append(ax.text(x=p_cents[0], y=bar_levels[0] + 0.2*space, s=p_text[0],
+                      size=label_size, ha='center'))
+    lines.append(l1)
+
+    if len(centers) > 2:
+        for idx, p_cent in enumerate(p_cents[:-1]):
+            ln = []
+            ln.append(ax.plot([centers[idx + 2]]*2,
+                              [tops[idx + 2] + space, bar_levels[idx + 1]],
+                              'k-'))
+            ln.append(ax.plot([centers[0], centers[idx + 2]],
+                              [bar_levels[idx + 1]]*2, 'k-'))
+            ln.append(ax.text(x=p_cents[idx + 1],
+                              y=bar_levels[idx + 1] + 0.2*space,
+                              s=p_text[idx+1], size=label_size, ha='center'))
+            lines.append(ln)
+
+    return lines
+
+
+def segment_colormap(cm_name, n_colors, n_pad=None, start=None):
+    """Segments a matplotlib colormap into descrete colors
+
+    Parameters
+    ----------
+    cm_name : str
+        The name of the continous matplotlib colormap which will be segmented.
+    n_colors : int
+        The number of colors needed in the colormap
+    n_pad : int, optional
+        The number of total colors to have in the colormap. By default, this
+        will be one more than the number of colors provided.
+    start : int, optional
+        The number of colors to skip over before display starts. By default,
+        this is `n_pad` - `n_colors`.
+
+    Returns
+    -------
+    new_map : array
+        A segmented array containing the colormap
+
+    """
+
+    # Sets parameters if necessary
+    if n_pad is None:
+        n_pad = n_colors + 1
+    if start is None:
+        start = n_pad - n_colors
+
+    # Gets the colormap
+    cm = mpl.cm.get_cmap(cm_name)
+    # Sets up the new map
+    new_map = np.array([list(cm(1. * (i + start) / n_pad)) for i in
+                        xrange(n_colors)])
+
+    return new_map
+
+
+def _get_bar_height(tops, factor=5):
+    """Calculates the lowest bar height"""
+    max_hi = tops.max()
+    # Gets the correct order of magnitdue
+    if max_hi < 1:
+        fudge = np.power(10, -np.floor(np.log(max_hi)))
+    else:
+        fudge = np.power(10, -np.ceil(np.log(max_hi)))
+    # Gets the correct rounding of the factor
+    if int(max_hi*fudge) < 5:
+        lowest = np.ceil(max_hi*fudge/factor)/(fudge)*factor
+    else:
+        lowest = np.ceil(max_hi*fudge*10/factor)/(fudge*10)*factor
+
+    return lowest, fudge
+
+
+def _get_p_value(sub_p, ref_group, group, p_tab_col):
+    """Determines the comparison value within a post-hoc table"""
+    if '%s vs. %s' % (ref_group, group) in sub_p['Group 2'].values:
```

Comment:
this can be simplified. given that it is used in an inner loop, this method should likely be optimized some. these if/elif blocks appear to be doing repeated linear time lookups (the `foo in blah.values`) as the `.values` is likely an `array` or `list`. from looking at the use of `_get_p_value`, it looks like `sub_p` varies outside of the forloop that this method is being called in. That suggests there may be some caching benefit. For instance, in the calling method, outside of the nested forloop this method is called in, something like this could be done:

```python
sub_p_lookups = {k: set(sub_p[k].values) for k in ('Group 1', 'Group 2')}
```

Then `_get_p_value` can be reduced some to:
```python
query_fwd = '%s vs. %s' % (ref_group, group)
query_rev = '%s vs. %s' % (group, ref_group)

for query, group in itertools.product([query_fwd, query_rev], ['Group 1', 'Group 2']):
    if query in sub_p_lookups[group]:
        return sub_p.loc[sub_p[group] == query, p_tab_col].values[0]
raise ValueError('%s is not a defined group' % query_fwd)
```